external/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -205,18 +205,28 @@ class AvroSerializer(rootCatalystType: DataType, rootAvroType: Schema, nullable:
       throw new IncompatibleSchemaException(s"Cannot convert Catalyst type $catalystStruct to " +
         s"Avro type $avroStruct.")
     }
-    val fieldConverters = catalystStruct.zip(avroStruct.getFields.asScala).map {
-      case (f1, f2) => newConverter(f1.dataType, resolveNullableType(f2.schema(), f1.nullable))
-    }
+
+    val (avroIndices: Array[Int], fieldConverters: Array[Converter]) =
+      catalystStruct.map { catalystField =>
+        val avroField = avroStruct.getField(catalystField.name)
+        if (avroField == null) {
```

Comment:
To fix `enum` issue, can we use the corresponding field at the same position when `avroField == null`?

Reply:
I think we should try matching the Avro field name under `union`s with `null`.

Reply:
When converting catalyst struct "a int, b int" -> avro type "b int, c int", we will update only the first avro field.

Reply:
When `avroField == null`, we should fail the writing because it's inconsistent between catalyst schema and avro schema.

Reply:
+1 for that.